abcpy/modelselections.py

```diff
@@ -0,0 +1,259 @@
+from abc import ABCMeta, abstractmethod
+import numpy as np
+from sklearn.ensemble import RandomForestClassifier, RandomForestRegressor
+
+
+
+class ModelSelections(metaclass = ABCMeta):
+    """This abstract base class defines a model selection rulr of how to choose model from a set of models
+    given an observation.
+ 
+            
+    """
+    
+    @abstractmethod
+    def __init__(self, model_array, statistics_calc, backend, seed = None):
+        """Constructor that must be overwritten by the sub-class.
+
+        The constructor of a sub-class must accept an array of models to choose the model 
+        from, and two non-optional parameters statistics calculator and backend stored in self.statistics_calc 
+        and self.backend defining how to calculate sumarry statistics from data and what kind of parallelization
+        to use.
+        
+        Parameters
+        ----------
+        model_array: list
+            A list of models which are of type abcpy.models.Model
+        statistics: abcpy.statistics.Statistics
+            Statistics object that conforms to the Statistics class.
+        backend: abcpy.backends.Backend
+            Backend object that conforms to the Backend class.
+        seed: integer, optional
+            Optional initial seed for the random number generator. The default value is generated randomly.    
+        """   
+        self.model_array = model_array
+        self.statistics_calc = statistics_calc
+        self.backend = backend
+        self.rng = np.random.RandomState(seed)
+        self.reference_table_calculated = None
+        
+
+        raise NotImplemented
+        
+        
+    @abstractmethod
+    def modelchoice(self, observations, n_samples = 1000, n_samplee_per_param = 100):
+        """To be overwritten by any sub-class: returns a model chosen by the modelselection
+        procedure most suitable to the obersved data set observations. It is assumed that observations is a 
+        list of n same type elements(eg., The observations can be a list containing n timeseries, n 
+        graphs or n np.ndarray). Further two optional integer arguments n_samples and n_samples_per_param
+        is supplied denoting the number of samples in the refernce table and the data points in each 
+        simulated data set.
+        
+        Parameters
+        ----------
+        data: python list
+            Contains n data sets.
+        n_samples : integer, optional
+            Number of samples to generate for reference table.
+        n_samples_per_param : integer, optional 
+            Number of data points in each simulated data set.  
+        Returns
+        -------
+        abcpy.models.Model
+            A model which are of type abcpy.models.Model
+            
+        """
+        
+        raise NotImplemented
+
+    @abstractmethod
+    def posteriorprobability(self, observations):
```

Comment:
The method would be more readable if we use posterior_probability. What do you think?